pal/src/sync/CCSpinLock.cpp

```diff
@@ -0,0 +1,114 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "pal/thread.hpp"
+#include "pal/cs.hpp"
+#include "pal_assert.h"
+
+#include <sched.h>
+#include <pthread.h>
+
+void CCSpinLock::Enter()
+{
+    size_t currentThreadId = -1, spin = 0;
+    if (shouldTrackThreadId)
+    {
+        currentThreadId = GetCurrentThreadId();
+        if (threadId == currentThreadId)
+        {
+            _ASSERTE(enterCount > 0);
+            enterCount++;
+            return;
+        }
+    }
+
+    while(__sync_lock_test_and_set(&lock, 1))
+    {
+        while(lock)
+        {
+            if (++spin > 11)
```

Comment:
small enough to let cache is updated. At the moment, this is best for the performance under high load. Adding a todo note for better tuning / parameterizing